src/random.h

```diff
@@ -227,13 +228,19 @@ class FastRandomContext
     template <typename Tp>
     Tp rand_uniform_delay(const Tp& time, typename Tp::duration range)
     {
-        using Dur = typename Tp::duration;
-        Dur dur{range.count() > 0 ? /* interval [0..range) */ Dur{randrange(range.count())} :
-                range.count() < 0 ? /* interval (range..0] */ -Dur{randrange(-range.count())} :
-                                    /* interval [0..0] */ Dur{0}};
-        return time + dur;
+        return time + rand_uniform_duration<Tp>(range);
     }
 
+    /** Generate a uniform random duration in the range from 0 (inclusive) to range (exclusive). */
+    template <typename Chrono>
+    typename Chrono::duration rand_uniform_duration(typename Chrono::duration range) noexcept
+    {
+        using Dur = typename Chrono::duration;
+        return range.count() > 0 ? /* interval [0..range) */ Dur{randrange(range.count())} :
+               range.count() < 0 ? /* interval (range..0] */ -Dur{randrange(-range.count())} :
```

Comment:
Some call sites may use a negative range. For example in addr relay, the timestamp of some address messages is randomly pushed into the past.

Also, it seems safer when the alternative is either unspecified behaviour, or a crash.

Reply:
Some call sites may use a negative range. For example in addr relay, the timestamp of some address messages is randomly pushed into the past.

Also, it seems safer when the alternative is either unspecified behaviour, or a crash.

Reply:
The behavior here still seems odd in those cases. I'd expect call sites that actually can result in negative range values to want "rand_uniform_duration(T)" to mean: "give me a uniform duration not exceeding T". If T is negative, that probably means just returning 0?

Reply:
Yeah, the interval is determined at run-time and the passed in value can either mean the start or the end of the interval, with the other value being `0`.

I am not sure if mapping any negative value to a constant `0` is less confusing. And I couldn't find another solution that is well defined at run time (no unspecified behaviour, no crash).

Reply:
Again, I am not sure that returning any constant is less confusing for a random function that took a non-zero input.

For reference, one potential call site would be:

```
addr.nTime = GetTime() - rng.randrange(nOneWeek) - nOneWeek;
```

This (and any other) call site can obviously be written without relying on a negative range, but we'd also need to avoid this with a compile error. Without `consteval`, this is probably only possible with some template magic?

Reply:
I don't think template magic works for this -- you get "non-type template parameter" errors trying to pass chrono::duration's as template parameters. I think you could do it with c++20 which allows "literal class types" as a non-type template parameter. https://en.cppreference.com/w/cpp/language/template_parameters

Reply:
Maybe `if (Assume(range.count()) > 0) { return Dur{randrange...}; } else { return range; }` ? I don't think we could make it a `consteval` function even with c++20, because `rng.randrange()` isn't going to be `consteval`?

Reply:
It would be ideal to enforce non-negative at compile time, and then force every caller to switch to a non-negative construction.

While I don't think how this is possible, I don't have much preference for the behavior if the value is negative. It is indeed annoying. We're not making it worse at least.
Perhaps status quo makes more sense in this case.

Reply:
Hmm, if I'm reading this right, this seems like quite an odd api to me. I'm not sure why passing a clock is necessary.

From what I can tell, you pass it a duration and a clock, and it gives you a new random duration in the clock's period. In other words.. we basically want a `std::chrono::duration_cast` which is hijacked and randomized. So why not use that api?

It'd look like (simplified for positive-only and untested):
```c++
template <class ToDuration, class Rep, class Period>
constexpr ToDuration rand_uniform_duration(const std::chrono::duration<Rep, Period>& d)
{
    auto max_ret = std::chrono::duration_cast<ToDuration>(d);
    auto rand_offset = ToDuration(randrange(d.count()));
    return max_ret - rand_offset;
}
```

Used like:
```c++
auto foo = rand_uniform_duration<std::chrono::milliseconds>(30s);
```

Or in this case:
```c++
auto foo = rand_uniform_duration<CThreadInterrupt::Clock::duration>(FEELER_SLEEP_WINDOW);

```

That way, we effectively communicate that we're randomizing in the target period as opposed to the provided one.

Reply:
I think a `duration_cast` is dangerous and not very useful here. Recall that the compiler will do the cast for you if it is safe to do. And in cases where it isn't safe, you are better off not calling this function in the first place.

I used clock, as it was suggested here: https://github.com/bitcoin/bitcoin/pull/24974#discussion_r861932886

Though, I think durations should be fine, too.

Reply:
As @ajtowns and I discussed yesterday in the office, here's a quick attempt at compile-time checking (to side-step the question of negative durations) just for fun. It's not completely unreasonable imo.

```c++
template <typename Duration, uint64_t Ticks>
struct static_duration
{
    using duration = Duration;
    static constexpr uint64_t ticks = Ticks;
};

template <class ToDuration, class static_duration>
constexpr ToDuration rand_uniform_duration(const static_duration& foo)
{
    typename static_duration::duration from(static_duration::ticks);
    auto max_ret = std::chrono::duration_cast<ToDuration>(from);
    auto rand_offset = ToDuration(randrange(max_ret.count()));
    return max_ret - rand_offset;
}

void test_func()
{
    static_duration<std::chrono::milliseconds, 30> foo;
    auto bar = rand_uniform_duration<std::chrono::milliseconds>(foo);
}
```

Reply:
Interesting. I guess you'd still have to static_assert against implicit sign conversion when going from 64-bit unsigned to 64-bit signed.

Reply:
Ok, here's a full working test-case. Statically checked for non-negative, passed durations, and no `duration_cast ` as @MarcoFalke suggested:

```c++
#include <chrono>
#include <cstdlib>
#include <cstdio>

long int randrange(uint64_t mod)
{
    auto ret = random() % mod;
    return ret;
}

template <typename Duration, uint64_t Ticks>
struct static_duration
{
    using duration = Duration;
    static constexpr uint64_t ticks = Ticks;
};

template <class ToDuration, class FromStaticDuration>
constexpr ToDuration rand_uniform_duration(const FromStaticDuration&)
{
    typename FromStaticDuration::duration from(FromStaticDuration::ticks);
    ToDuration max_ret{from};
    auto rand_offset = ToDuration(randrange(max_ret.count()));
    return max_ret - rand_offset;
}

int main()
{
    srand(time(nullptr));
    static constexpr static_duration<std::chrono::seconds, 30> foo;
    auto bar = rand_uniform_duration<std::chrono::milliseconds>(foo);
    printf("duration: %ld\n", bar.count());
}
```

Reply:
> Interesting. I guess you'd still have to static_assert against implicit sign conversion when going from 64-bit unsigned to 64-bit signed.

Right, I guess you'd want to accept a `int64_t` and `static_assert` that it's > 0 instead.

Feel free to disregard or clean up and use :)

Reply:
Ok, glad we figured out how to implement this, but is this something that is actually preferable? :sweat_smile:

Reply:
🤷
It's perfectly expressive and constexpr, so there's not much downside other than (the syntax and) the emitted per-constant symbol. The obvious realistic downside being that it's kinda overkill...

But maybe not it's a bad thing to be able to express only-positive-time more generally? So maybe that's the test... is there somewhere else where it might be useful to describe time constants as positive only?

Reply:
I'm happy (or about equally unhappy...) with any of the approaches really; and I guess we'd want to switch to C++20 syntax when that's available either way, so *shrug* from me. @sipa?

Reply:
I think there is a *minimal, theoretical* advantage to disallowing negative values, as the inverting the most negative value is an integer sanitizer violation. Though, this shouldn't affect execution of a program that is not compiled with the integer sanitizer and shouldn't happen in practise either.

What is the "C++20" syntax, if I may ask?

Reply:
With C++20 you can pass a duration to a template, and then do a static_assert on its value, so you'd write `rand_uniform_duration<dur>()` instead of `rand_uniform_duration(dur)` and get the compile-time check.

Reply:
Oh, with `dur` you mean `static_duration`, not `std::chrono::duration`?

I can't get it to work with `std::chrono::duration`:

```cpp
#include <chrono>

using namespace std::chrono_literals;

struct FRC {
    template <std::chrono::duration range>
    auto rand_uniform_delay(auto time) {
        return time + rand_uniform_duration<decltype(time)::duration, range>();
    }
    template <typename Duration, std::chrono::duration range>
    Duration rand_uniform_duration() noexcept {
        static_assert(range.count() > 0);
        return Duration{randrange(range.count())};
    }
};

int main() {
    FRC frc;
    frc.rand_uniform_delay<9h>(std::chrono::sys_seconds{1s});
    frc.rand_uniform_duration<std::chrono::seconds, 9h>();
}
```

https://godbolt.org/z/8Y868s98K

Reply:
Happy to use different std-lib syntax, but anything that requires a wrapper class that re-implements stuff seems overkill to solve a problem that doesn't exist.